File: 998数据结构/001递归/mergeSort.py
a = [3,5,9,20,1,7,67,6,8,4,2]

def _merge(array, start, middle, end):
    len1 = middle - start + 1
    len2 = end - middle
    array1 = [0] * len1
    array2 = [0] * len2

    for i in range(len1):
        array1[i] = array[start + i]

    for j in range(len2):
        array2[j] = array[middle + j + 1]

    i = 0
    j = 0
    for k in range(start, end+1):
       if i==len1 or (j!=len2 and array2[j] < array1[i]):
            array[k] = array2[j]
            j = j + 1
       else:
            array[k] = array1[i]
            i = i + 1


def _merge_sort(array, start, end):
    if start >= end:
       return
    else :
        middle = int((start + end)/2)
        _merge_sort(array, start, middle) # middle as the end
        _merge_sort(array, middle+1, end)
        _merge(array, start, middle, end)


def merge_sort(array):
	_merge_sort(array, 0, len(array)-1)


merge_sort(a)
print(a)
    

File: 997Algorithm/2_Introduction_to_Algorithms/merge_sort.cpp
#include <stdio.h>
#include <string.h>
#include <algorithm>
#include "../include/algo_util.h"

const int len = 10;
int a[len];
int b[len];

void merge(int arr[], int s, int m, int e){
    int len1 = m - s + 1;
    int len2 = e - m;
    int brr[len1];
    int crr[len2];
    for (int i=0; i<len1; i++)
        brr[i] = arr[i+s];
    for (int i=0; i<len2; i++)
        crr[i] = arr[i+m+1];
    int i=0, j=0;
    for (int k=s; k<=e; k++){
        if (i<len1 && j<len2) {
            if (brr[i] > crr[j])
                arr[k] = crr[j++];
            else
                arr[k] = brr[i++];
        } else if (i == len1)
            arr[k] = crr[j++];
        else
            arr[k] = brr[i++];
    }
}

void _merge_sort(int arr[], int s, int e){
    if(s<e) {
        int m = (e+s)/2;
        _merge_sort(arr, s, m);
        _merge_sort(arr, m+1, e);
        merge(arr, s, m, e);
    }
}

void merge_sort(int arr[], int s, int e){
    _merge_sort(arr, s, e);
}

int main(){
    /*
    int a[] = {1,2};
    print_arr(a,2);
    merge(a,0,0,1);
    print_arr(a,2);
    */
    
    init_rand_ints(a, len);
    memcpy(b, a, len*sizeof(int));
    std::sort(b, b+len);
    merge_sort(a, 0, len-1);
    if (is_equal(a, b, len))
        printf("ok\n");
    else
        printf("no!\n");
    return 0;
}
